### PR TITLE
gh-95051: ensure that timeouts scheduled with `asyncio.Timeout` that have already expired are deliverered promptly

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-95051

### CODE REVIEW

File: Lib/test/test_asyncio/test_timeouts.py [hunk 1]
```diff
@@ -103,7 +103,31 @@ async def test_timeout_zero(self):
         self.assertTrue(cm.expired())
         # 2 sec for slow CI boxes
         self.assertLess(t1-t0, 2)
-        self.assertTrue(t0 <= cm.when() <= t1)
+        self.assertEqual(cm.when(), 0)
```
Comment (graingert) [inline Lib/test/test_asyncio/test_timeouts.py]:
```suggestion
        self.assertTrue(t0 <= cm.when() <= t1)
```

File: Lib/test/test_asyncio/test_timeouts.py [hunk 2]
```diff
@@ -105,6 +105,30 @@ async def test_timeout_zero(self):
...
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(0) as cm:
+                await asyncio.sleep(0)
+        t1 = loop.time()
+        self.assertTrue(cm.expired())
+        # 2 sec for slow CI boxes
+        self.assertLess(t1-t0, 2)
+        self.assertEqual(cm.when(), 0)
```
Comment (graingert) [inline Lib/test/test_asyncio/test_timeouts.py]:
```suggestion
        self.assertTrue(t0 <= cm.when() <= t1)
```

File: Lib/test/test_asyncio/test_timeouts.py [hunk 3]
```diff
@@ -105,6 +105,30 @@ async def test_timeout_zero(self):
...
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(-11) as cm:
+                await asyncio.sleep(0)
+        t1 = loop.time()
+        self.assertTrue(cm.expired())
+        # 2 sec for slow CI boxes
+        self.assertLess(t1-t0, 2)
+        self.assertEqual(cm.when(), 0)
```
Comment (graingert) [inline Lib/test/test_asyncio/test_timeouts.py]:
```suggestion
        self.assertTrue(t0 <= cm.when() <= t1)
```

File: Lib/asyncio/timeouts.py [hunk 4]
```diff
@@ -126,7 +126,13 @@ def timeout(delay: Optional[float]) -> Timeout:
...
     """
     loop = events.get_running_loop()
-    return Timeout(loop.time() + delay if delay is not None else None)
+    if delay is None:
+        return Timeout(None)
+
+    if delay <= 0:
+        return Timeout(0)
+
+    return Timeout(loop.time() + delay)
```
Comment (graingert) [inline Lib/asyncio/timeouts.py]:
```suggestion
    return Timeout(loop.time() + delay if delay is not None else None)
```

File: Misc/NEWS.d/next/Library/2022-07-21-22-59-22.gh-issue-95109.usxA9r.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+special case asyncio.timeout(0) and timeouts in the past
```
Comment (graingert) [inline Misc/NEWS.d/next/Library/2022-07-21-22-59-22.gh-issue-95109.usxA9r.rst]:
```suggestion
ensure that timeouts scheduled with :class:`asyncio.Timeout` that have already expired are delivered promptly
```

File: Lib/test/test_asyncio/test_timeouts.py [hunk 6]
```diff
@@ -105,6 +105,30 @@ async def test_timeout_zero(self):
...
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(-11) as cm:
+                await asyncio.sleep(0)
+        t1 = loop.time()
+        self.assertTrue(cm.expired())
+        # 2 sec for slow CI boxes
+        self.assertLess(t1-t0, 2)
+        self.assertTrue(t0 <= cm.when() <= t1)
```
Comment (graingert) [inline Lib/test/test_asyncio/test_timeouts.py]:
```suggestion
        self.assertTrue(t0 >= cm.when() <= t1)
```

File: Misc/NEWS.d/next/Library/2022-07-21-22-59-22.gh-issue-95109.usxA9r.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+ensure that timeouts scheduled with :class:`asyncio.Timeout` that have already expired are delivered promptly
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Library/2022-07-21-22-59-22.gh-issue-95109.usxA9r.rst]:
```suggestion
Ensure that timeouts scheduled with :class:`asyncio.Timeout` that have already expired are delivered promptly.
```

File: Lib/asyncio/timeouts.py [hunk 8]
```diff
@@ -52,10 +52,10 @@ def reschedule(self, when: Optional[float]) -> None:
             self._timeout_handler = None
         else:
             loop = events.get_running_loop()
-            self._timeout_handler = loop.call_at(
-                when,
-                self._on_timeout,
-            )
+            if loop.time() >= when:
```
Comment (gvanrossum) [inline Lib/asyncio/timeouts.py]: I don't know if it's just me, but I find it reads better if we reverse the condition, like so:
```suggestion
            if when <= loop.time():
```

**Review by gvanrossum — COMMENTED** [review]
LG, just a nit, and maybe now that we're testing this behavior, it should be noted in the docs?

**Review by gvanrossum — APPROVED** [review]

**Comment by gvanrossum** [comment]
@graingert please backport manually and ask Pablo for a review.

**Comment by AlexWaygood** [comment]
> @graingert please backport manually and ask Pablo for a review.

When the message is "I had trouble checking out the 3.X branch" (as opposed to "I couldn't create the backport due to a merge conflict"), just removing and re-adding the label is usually enough to fix it :)

**Comment by AlexWaygood** [comment]
> [GH-95216](https://github.com/python/cpython/pull/95216) is a backport of this pull request to the [3.11 branch](https://github.com/python/cpython/tree/3.11).

@graingert, I already removed and re-added the label — this is the backport here

**Comment by graingert** [comment]
argh I saw the email notification and had a stale copy of the PR page open

**Comment by AlexWaygood** [comment]
> argh I saw the email notification and had a stale copy of the PR page open

It's race conditions all the way down

**Comment by gvanrossum** [comment]
Maybe one of you can make the bot more user friendly?

**Comment by graingert** [comment]
It's race conditions all the way down

**Comment by AlexWaygood** [comment]
> Maybe one of you can make the bot more user friendly?

It's been on my to-do list for a while.....

**Pull request merged**